cmd/atlasci/internal/ci/run.go

```diff
@@ -0,0 +1,64 @@
+// Copyright 2021-present The Atlas Authors. All rights reserved.
+// This source code is licensed under the Apache 2.0 license found
+// in the LICENSE file in the root directory of this source tree.
+
+package ci
+
+import (
+	"context"
+
+	"ariga.io/atlas/sql/migrate"
+	"ariga.io/atlas/sql/sqlcheck"
+	"ariga.io/atlas/sql/sqlclient"
+)
+
+// Runner is used to execute CI jobs.
+type Runner struct {
+	// DevClient configures the "dev driver" to calculate
+	// migration changes by the driver.
+	Dev *sqlclient.Client
+
+	// RunChangeDetector configures the ChangeDetector to
+	// be used by the runner.
+	ChangeDetector ChangeDetector
+
+	// Scan is used for scanning the migration directory.
+	Scan migrate.Scanner
+
+	// Analyzer defines the analysis to be run in the CI job.
+	Analyzer sqlcheck.Analyzer
+
+	// Reporter is used to report diagnostics in the CI.
+	Reporter sqlcheck.Reporter
+}
+
+// Run executes the CI job.
+func (r *Runner) Run(ctx context.Context) error {
+	base, feat, err := r.ChangeDetector.DetectChanges(ctx)
+	if err != nil {
+		return err
+	}
+	// Bring the dev environment to the base point.
+	for _, f := range base {
+		stmt, err := r.Scan.Stmts(f)
+		if err != nil {
+			return err
+		}
+		for _, s := range stmt {
+			if _, err := r.Dev.ExecContext(ctx, s); err != nil {
+				return err
+			}
+		}
+	}
```

Comment:
Why not use an `migrate.Executor` here? The Executor will be able (soon) to e.g. check for checkpoints and stuff to speed up reaching base. By doing it the way you do, we have to built the logic twice (once we do).

There is a `NoopRevisionReadWriter` for exactly that case. 

For the `Dir` / `Scanner` you can use an "in memory" implementation to serve the base files.

Reply:
I missed the `NoopRevisionReadWriter` one. I'll give it a look now.

Reply:
I see now that it depends on the hash file. Let's talk about it offline.